Update all public API javadoc

tomblench opened the pull request:
See #400

Note that the query interface is liable to change soon so any documentation changes here are moot.

- Ensure method descriptions are correct
- Fix formatting
- Add target="_blank" for href links
- Fix throws clauses
- (not documentation) Tidy up definition of custom Exceptions

emlaver commented:
add period here: https://github.com/cloudant/sync-android/pull/446/files#diff-52579adb2b746a1d2c45b644deca59ffR360

emlaver commented:
Is it worth mentioning `OperationStartedListener.operationStarted` here: https://github.com/cloudant/sync-android/pull/446/files#diff-52579adb2b746a1d2c45b644deca59ffR359

emlaver commented on cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/documentstore/Attachment.java:
```diff
@@ -22,7 +22,13 @@
 /**
- * Base class for Attachments
+ * <p>
+ *   An attachment associated with a CouchDB document
```
Add period.

emlaver commented on cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/documentstore/AttachmentNotSavedException.java:
```diff
@@ -19,19 +19,29 @@
+ *   Thrown when there is an error saving attachments to local storage.
+ * </p>
+ * <p>
+ *   Likely causes of this exception include: IO exception whilst saving attachment to local
```
What about using `<li>` to list the exceptions?

emlaver commented on cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/documentstore/ConflictResolver.java:
```diff
@@ -17,31 +17,35 @@
 
 /**
- * Interface to implement conflicts resolving algorithm.
+ * The {@code ConflictResolver} interface should be implemented by classes which are used to resolve
```
What about `...interface should be implemented by classes which resolve conflicts via...`

emlaver commented:
Some of the javadoc for @param and @throws ends in a period.  I think it should either end in a period or not to keep things consistent.

emlaver commented on cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/documentstore/Database.java:
```diff
@@ -238,7 +244,7 @@ DocumentRevision read(String documentId, String revisionId) throws
      * @param resolver the ConflictResolver used to resolve
      *                 conflicts
-     * @throws ConflictException Found new conflicts while
+     * @throws ConflictException If new conflicts were found while
```
lowercase: `if new conflicts...`

emlaver commented on cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/documentstore/DocumentException.java:
```diff
@@ -20,24 +20,36 @@
+ * {@link DocumentRevision}s or their associated {@link Attachment}s
+ * </p>
+ * <p>
+ * Note that these error cases are distinct from {@link DocumentStoreException}s which are thrown
```
`Note: these error...`

emlaver commented on cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/documentstore/DocumentException.java:
```diff
@@ -20,24 +20,36 @@
+ * when an unexpected condition was encountered (for example, an internal SQLite database error).
+ * </p>
+ * <p>
+ * This is the base class of a hierarchy of exceptions: in most cases a more specific
```
Is the colon needed? `...hierarchy of exceptions.  In most cases, a more...`

emlaver commented on cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/documentstore/ConflictResolver.java:
```diff
@@ -17,31 +17,35 @@
      *
-     * @param docId id of the Document with conflicts
-     * @param conflicts list of conflicted DocumentRevision, including
+     * @param docId id of the {@link DocumentRevision} with conflicts
```
capitalize: `ID of..`

emlaver commented on cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/documentstore/Database.java:
```diff
@@ -238,7 +244,7 @@ DocumentRevision read(String documentId, String revisionId) throws
      * @param docId id of Document to resolve conflicts
```
capitalize: `ID of..`

emlaver commented on cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/documentstore/Database.java:
```diff
@@ -264,7 +270,7 @@ void resolveConflicts(String docId, ConflictResolver resolver)
      * @throws AttachmentException if there was an error saving any new attachments
      * @throws InvalidDocumentException if the document body was invalid
      * @throws ConflictException if a document with this document id already exists
```
capitalize: `ID of..`

emlaver commented on cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/documentstore/DocumentNotFoundException.java:
```diff
@@ -21,41 +21,38 @@
      * Creates a document not found exception with the default message
+     *
      * @param docId The document id of the document that could not be found
      * @param revId The rev id of the document that could not be found
```
capitalize both document and rev ID.

emlaver commented on cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/documentstore/DocumentStoreException.java:
```diff
@@ -16,9 +16,18 @@
+
+/**
+ * <p>
+ * Thrown when an unexpected condition was encountered during a {@link DocumentStore} method (for
```
Remove parenthesis: `...method, for example an internal...`

emlaver commented on cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/documentstore/DocumentStoreNotOpenedException.java:
```diff
@@ -15,7 +15,8 @@
 /**
- * Exception thrown when an existing datastore cannot be opened or a new datastore cannot be created
+ * Trown when an existing {@link DocumentStore} cannot be opened or a new {@link DocumentStore}
+ * cannot be created
```
missing period.

emlaver commented on cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/documentstore/InvalidDocumentException.java:
```diff
@@ -19,11 +19,12 @@
 
-    }
+/**
+ * Thrown when a {@link DocumentRevision} contains errors: for example if it contains top-level
```
Change colon to something like `...contains errors.  For example, if`

emlaver commented on cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/documentstore/LocalDocument.java:
```diff
@@ -15,7 +15,7 @@
 
 /**
- * A local Document, this document does not have a history, or the concept of revisions
+ * A local Document. {@code LocalDocument}s do not have a history, or the concept of revisions
```
Did you want to use the @link here?

tomblench replied:
I'm in two minds about that, Android Studio gives a warning if documentation links to itself, but the advantage of using a link is that it would catch any typos in the link name

emlaver commented on cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/notifications/DocumentModified.java:
```diff
@@ -17,6 +17,8 @@
 import com.cloudant.sync.documentstore.DocumentRevision;
 
 /**
+ * Generic event for document create/update/delete
```
`...document create, update, and delete.`

emlaver commented on cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/notifications/DocumentStoreModified.java:
```diff
@@ -15,6 +15,8 @@
 package com.cloudant.sync.event.notifications;
 
 /**
+ * Generic event for database create/delete
```
`...database create or delete`

emlaver commented on cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/query/QueryException.java:
```diff
@@ -17,6 +17,10 @@
 /**
+ * <p>
+ * Thrown when an unexpected condition was encountered during a {@link Query} method (for
+ * example, an internal SQLite database error).
```
Use a comma instead of parenthesis.

tomblench commented:
@emlaver I think I've addressed most comments except for the issue of capitalisation and punctuation.

In the Java API javadoc, they [sometimes](https://docs.oracle.com/javase/7/docs/api/java/util/Formatter.html#format(java.util.Locale,%20java.lang.String,%20java.lang.Object...)) use a capital initial letter and full stop for parameters, and sometimes [don't](https://docs.oracle.com/javase/7/docs/api/java/util/HashMap.html#put(K,%20V))

I agree that we ought to be consistent on this but I don't think it's a blocker and doesn't harm the clarity of the documentation. It would also take a long time do this for all of our javadoc comments

emlaver commented:
Ok, this makes sense and I agree that it doesn't harm the clarify of the java docs.
